swarm/tools/selftest.py

```
@@ -577,9 +577,11 @@ def run_step(self, step: SelfTestStep, all_results: Dict[str, SelfTestResult]) -
         try:
             # Use Popen with start_new_session=True to create a new process group.
             # This ensures we can kill all child processes on timeout, not just the shell.
+            # Parse command string into list for safe execution (avoids shell injection)
+            cmd_args = shlex.split(step.full_command())
```

Ruff (0.14.13)

Undefined name `shlex`

(F821)